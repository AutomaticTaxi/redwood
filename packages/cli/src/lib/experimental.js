import path from 'path'

import { getConfig, getPaths, resolveFile } from '@redwoodjs/project-config'

/**
 * Checks to see if the user has the experimental server file present at api/src/server.ts|js and enabled in redwood.toml
 *
 * @returns {boolean} true if the experimental server file is enabled and present
 */
export function isUsingExperimentalServerFile() {
  return (
    getConfig().experimental?.serverFile?.enabled &&
    resolveFile(path.join(getPaths().api.src, 'server'))
  )
}
