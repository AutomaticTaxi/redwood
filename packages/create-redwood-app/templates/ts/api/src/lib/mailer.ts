import { Mailer } from '@redwoodjs/mailer-core'
import { InMemoryMailHandler } from '@redwoodjs/mailer-handler-in-memory'
import { NodemailerMailHandler } from '@redwoodjs/mailer-handler-nodemailer'
import { StudioMailHandler } from '@redwoodjs/mailer-handler-studio'
import { ReactEmailRenderer } from '@redwoodjs/mailer-renderer-react-email'

import { logger } from './logger'

export const mailer = new Mailer({
  // Handlers to send our emails out to the world
  handling: {
    handlers: {
      inMemory: new InMemoryMailHandler(),
      studio: new StudioMailHandler(),
      // TODO: Update this handler config or switch it out for a different handler completely
      nodemailer: new NodemailerMailHandler({
        transport: {
          host: 'localhost',
          port: 4319,
          secure: false,
        },
      }),
    },
    // This is our default handler when in production only
    default: 'nodemailer',
  },

  // Renderers to convert our templates into HTML/text
  rendering: {
    renderers: {
      reactEmail: new ReactEmailRenderer(),
    },
    default: 'reactEmail',
  },

  // Specific handlers for testing and development time
  test: {
    handler: 'inMemory',
    // when: process.env.NODE_ENV === 'test',
  },
  development: {
    handler: 'studio',
    // when: process.env.NODE_ENV !== 'production',
  },

  // Default send options
  defaults: {
    // replyTo: 'noreply@example.com',
  },

  logger,
})
