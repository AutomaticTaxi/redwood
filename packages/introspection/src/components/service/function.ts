import path from 'node:path'

import type { ExportNamedDeclaration } from '@babel/types'
import {
  isArrowFunctionExpression,
  isFunctionExpression,
  isIdentifier,
  isObjectPattern,
  isObjectProperty,
  isVariableDeclaration,
  isVariableDeclarator,
} from '@babel/types'

import type {
  RedwoodIntrospectionError,
  RedwoodIntrospectionWarning,
} from '../introspection'
import { RedwoodIntrospectionComponent } from '../introspection'

export class RedwoodServiceFunction extends RedwoodIntrospectionComponent {
  readonly type = 'service-function'

  private readonly constructorErrors: RedwoodIntrospectionError[] = []

  readonly serviceName: string
  readonly parameters: string[] = []

  private constructor(
    filepath: string,
    exportNamedDeclaration: ExportNamedDeclaration
  ) {
    // Extract the function name
    let name: string | undefined = undefined
    if (
      isVariableDeclaration(exportNamedDeclaration.declaration) &&
      exportNamedDeclaration.declaration.declarations[0] != null &&
      isVariableDeclarator(
        exportNamedDeclaration.declaration.declarations[0]
      ) &&
      isIdentifier(exportNamedDeclaration.declaration.declarations[0].id)
    ) {
      name = exportNamedDeclaration.declaration.declarations[0].id.name
    }

    super(filepath, name)
    this.serviceName = path.parse(this.filepath).name

    if (name === undefined) {
      this.constructorErrors.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: 'Unable to extract the name of the service function',
      })
      this.constructorErrors.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: 'Unable to extract the parameters of the service function',
      })
      return
    }

    // Extract the function parameters
    if (
      isVariableDeclaration(exportNamedDeclaration.declaration) &&
      exportNamedDeclaration.declaration.declarations[0] != null &&
      isVariableDeclarator(
        exportNamedDeclaration.declaration.declarations[0]
      ) &&
      (isArrowFunctionExpression(
        exportNamedDeclaration.declaration.declarations[0].init
      ) ||
        isFunctionExpression(
          exportNamedDeclaration.declaration.declarations[0].init
        ))
    ) {
      const parameters =
        exportNamedDeclaration.declaration.declarations[0].init.params[0]
      if (isObjectPattern(parameters)) {
        parameters.properties.forEach((property) => {
          if (isObjectProperty(property) && isIdentifier(property.key)) {
            this.parameters.push(property.key.name)
          }
        })
      }
    }
  }

  // executeAdditionalChecks() {
  // const associatedSDLOperation = this.getAssociatedSDLOperation()
  // if (associatedSDLOperation) {
  //   let parametersMatch =
  //     associatedSDLOperation.parameters.length === this.parameters.length
  //   for (let i = 0; i < this.parameters.length; i++) {
  //     parametersMatch &&=
  //       this.parameters[i] === associatedSDLOperation.parameters[i]
  //   }
  //   if (!parametersMatch) {
  //     this.errors.push({
  //       code: RedwoodErrorCode.SERVICE_FUNCTION_PARAMETERS_DO_NOT_MATCH_SDL_OPERATION,
  //       message: `The parameters of the "${this.name}" function do not match those of the correspondind SDL operation "${associatedSDLOperation.name}"`,
  //     })
  //   }
  // }
  // }

  getErrors(): RedwoodIntrospectionError[] {
    const errors: RedwoodIntrospectionError[] = [...this.constructorErrors]
    return errors
  }

  getWarnings(): RedwoodIntrospectionWarning[] {
    // No warnings for services
    return []
  }

  static parseServiceFunction(
    filepath: string,
    exportNamedDeclaration: ExportNamedDeclaration
  ) {
    return new RedwoodServiceFunction(filepath, exportNamedDeclaration)
  }

  // getAssociatedSDLOperation() {
  //   const sdl = RedwoodProject.getProject({ pathWithinProject: this.filepath })
  //     .getSDLs()
  //     .find((sdl) => {
  //       return sdl.name === this.serviceName
  //     })
  //   if (sdl) {
  //     const query = sdl.queries?.find((query) => {
  //       return query.name === this.name
  //     })
  //     const mutation = sdl.mutations?.find((mutation) => {
  //       return mutation.name === this.name
  //     })
  //     return query || mutation
  //   }
  //   return undefined
  // }
}
