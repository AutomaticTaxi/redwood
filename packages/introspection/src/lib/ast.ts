import fs from 'node:fs'
import path from 'node:path'

import { parse as babelParse } from '@babel/parser'
import type { ParserPlugin } from '@babel/parser'
import traverse from '@babel/traverse'
import type {
  JSXElement,
  ExportNamedDeclaration,
  File,
  VariableDeclarator,
  ExportDefaultDeclaration,
} from '@babel/types'
import {
  assertJSXOpeningElement,
  assertJSXIdentifier,
  assertIdentifier,
  isJSXAttribute,
  isJSXSpreadAttribute,
  isVariableDeclaration,
  isVariableDeclarator,
  isIdentifier,
} from '@babel/types'

import { getPaths } from '@redwoodjs/project-config'

export function getASTFromFile(filepath: string) {
  const code = fs.readFileSync(filepath, { encoding: 'utf8', flag: 'r' })

  // use jsx plugin for web files, because in JS, the .jsx extension is not used
  const isJsxFile =
    path.extname(filepath).match(/[jt]sx$/) ||
    path.parse(filepath).dir.startsWith(getPaths().web.base)

  const plugins = [
    'typescript',
    'nullishCoalescingOperator',
    'objectRestSpread',
    isJsxFile && 'jsx',
  ].filter(Boolean) as ParserPlugin[]

  return babelParse(code, {
    sourceType: 'module',
    plugins,
  })
}

export function getSpecificNamedExportDeclarations(ast: File, names: string[]) {
  const desiredNamedExports: Map<string, VariableDeclarator> = new Map()
  const allNamedExports: ExportNamedDeclaration[] = []
  traverse(ast, {
    ExportNamedDeclaration: (path) => {
      allNamedExports.push(path.node)
    },
  })
  allNamedExports.forEach((namedExport) => {
    if (
      namedExport.declaration != null &&
      isVariableDeclaration(namedExport.declaration) &&
      isVariableDeclarator(namedExport.declaration.declarations[0]) &&
      isIdentifier(namedExport.declaration.declarations[0].id) &&
      names.includes(namedExport.declaration.declarations[0].id.name)
    ) {
      desiredNamedExports.set(
        namedExport.declaration.declarations[0].id.name,
        namedExport.declaration.declarations[0]
      )
    }
  })
  return desiredNamedExports
}

export function getExportDefaultDeclaration(ast: File) {
  let defaultExport: ExportDefaultDeclaration | undefined
  traverse(ast, {
    ExportDefaultDeclaration: (path) => {
      defaultExport = path.node
    },
  })
  return defaultExport
}

export function getVariableDeclarator(ast: File, name: string) {
  let variableDeclarator: VariableDeclarator | undefined
  traverse(ast, {
    VariableDeclarator: (path) => {
      if (isIdentifier(path.node.id) && path.node.id.name === name) {
        variableDeclarator = path.node
      }
    },
  })
  return variableDeclarator
}

export function getJSXElementAttributes(element: JSXElement) {
  const attributes: Map<string, string> = new Map()
  assertJSXOpeningElement(element.openingElement)
  if (element.openingElement.attributes.length > 0) {
    element.openingElement.attributes.forEach((attribute) => {
      if (isJSXAttribute(attribute)) {
        assertJSXIdentifier(attribute.name)
        const key = attribute.name.name
        let value: string
        switch (attribute.value?.type) {
          case 'StringLiteral':
            value = attribute.value.value
            break
          case 'JSXExpressionContainer':
            assertIdentifier(attribute.value.expression)
            value = attribute.value.expression.name
            break
          case undefined: // Accept this because it could be a attribute with no value, i.e. notfound
            value = 'true'
            break
          default:
            throw new Error(
              `Unknown value type found while trying to process the ${key} property when extracting all attributes`
            )
        }
        attributes.set(key, value)
      } else if (isJSXSpreadAttribute(attribute)) {
        throw new Error(
          'JSXSpreadAttribute type is not yet supported when extracting all attributes'
        )
      } else {
        throw new Error(
          'Uknown attribute type found while trying to extract all attributes'
        )
      }
    })
  }
  return attributes
}
