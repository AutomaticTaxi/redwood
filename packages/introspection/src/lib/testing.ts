import { ensurePosixPath } from '@redwoodjs/project-config'

export function stripAndFormatPathForTesting(
  filepath: string,
  projectFilepath: string
) {
  if (filepath.startsWith(projectFilepath)) {
    filepath = ensurePosixPath(filepath.substring(projectFilepath.length))
  }
  return filepath
}
