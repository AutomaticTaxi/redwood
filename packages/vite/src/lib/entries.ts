import path from 'node:path'

import type { PagesDependency } from '@redwoodjs/project-config'
import {
  ensurePosixPath,
  getPaths,
  processPagesDir,
} from '@redwoodjs/project-config'

const getPathRelativeToSrc = (maybeAbsolutePath: string) => {
  // If the path is already relative
  if (!path.isAbsolute(maybeAbsolutePath)) {
    return maybeAbsolutePath
  }

  return `./${path.relative(getPaths().web.src, maybeAbsolutePath)}`
}

const withRelativeImports = (page: PagesDependency) => {
  return {
    ...page,
    relativeImport: ensurePosixPath(getPathRelativeToSrc(page.importPath)),
  }
}

export function getEntries() {
  return processPagesDir()
    .map(withRelativeImports)
    .reduce(
      (acc, page) => {
        acc[page.importName] = page.path
        return acc
      },
      {} as Record<string, string>,
    )
}

export async function getEntriesFromDist(): Promise<Record<string, string>> {
  const entriesDist = getPaths().web.distRscEntries
  const { serverEntries } = await import(`file://${entriesDist}`)
  return serverEntries
}
