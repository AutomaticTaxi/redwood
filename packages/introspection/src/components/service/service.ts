import fs from 'node:fs'
import path from 'node:path'

import traverse from '@babel/traverse'

import { getPaths } from '@redwoodjs/project-config'

import { getASTFromFile } from '../../lib/ast'
import type {
  RedwoodIntrospectionError,
  RedwoodIntrospectionWarning,
} from '../introspection'
import { RedwoodIntrospectionComponent } from '../introspection'

import { RedwoodServiceFunction } from './function'

export class RedwoodService extends RedwoodIntrospectionComponent {
  readonly type = 'service'

  readonly functions: RedwoodServiceFunction[] = []

  private constructor(filepath: string) {
    super(filepath)
    const ast = getASTFromFile(filepath)
    traverse(ast, {
      ExportNamedDeclaration: (path) => {
        this.functions.push(
          RedwoodServiceFunction.parseServiceFunction(this.filepath, path.node)
        )
      },
    })
  }

  getErrors(): RedwoodIntrospectionError[] {
    // No errors for services
    return []
  }

  getWarnings(): RedwoodIntrospectionWarning[] {
    // No warnings for services
    return []
  }

  static parseService(filepath: string) {
    return new RedwoodService(filepath)
  }

  static parseServices(directory: string = getPaths().api.services) {
    const services: RedwoodService[] = []

    // TODO: Confirm this is the condition to detect a service
    const getServiceFiles = (directory: string) => {
      const serviceFiles: string[] = []
      const directoryContents = fs.readdirSync(directory)
      directoryContents.forEach((content) => {
        const stat = fs.lstatSync(path.join(directory, content))
        if (stat.isDirectory()) {
          serviceFiles.push(...getServiceFiles(path.join(directory, content)))
        } else if (stat.isFile()) {
          if (
            content.match(/.+\.(js|ts)$/) &&
            !content.match(/.+\.(test\.js|test\.ts)$/) &&
            !content.match(/.+\.(scenarios\.js|scenarios\.ts)$/)
          ) {
            serviceFiles.push(path.join(directory, content))
          }
        }
      })
      return serviceFiles
    }

    const serviceFiles = getServiceFiles(directory)
    serviceFiles.forEach((servicePath) => {
      services.push(this.parseService(servicePath))
    })

    return services
  }
}
