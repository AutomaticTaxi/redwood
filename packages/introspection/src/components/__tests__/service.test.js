import path from 'node:path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { RedwoodService } from '../service/service'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)
const FIXTURE_PROJECTS = [
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
]

describe.each(FIXTURE_PROJECTS)(
  "From within the '%s' fixture",
  (PROJECT_NAME) => {
    // Setup RWJS_CWD to point to the fixture project
    const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)
    const RWJS_CWD = process.env.RWJS_CWD
    beforeAll(() => {
      process.env.RWJS_CWD = PROJECT_PATH
    })
    afterAll(() => {
      process.env.RWJS_CWD = RWJS_CWD
    })

    it('returns the correct services', () => {
      const services = RedwoodService.parseServices()
      expect(services.length).toMatchSnapshot('services count')
      for (let i = 0; i < services.length; i++) {
        const service = services[i]
        service.filepath = stripAndFormatPathForTesting(
          service.filepath,
          PROJECT_PATH
        )
        for (let j = 0; j < service.functions.length; j++) {
          service.functions[j].filepath = stripAndFormatPathForTesting(
            service.functions[j].filepath,
            PROJECT_PATH
          )
        }

        expect(service).toMatchSnapshot(`${service.filepath} content`)
        expect(service.getErrors()).toMatchSnapshot(
          `${service.filepath} errors`
        )
        expect(service.getWarnings()).toMatchSnapshot(
          `${service.filepath} warnings`
        )
      }
    })
  }
)
