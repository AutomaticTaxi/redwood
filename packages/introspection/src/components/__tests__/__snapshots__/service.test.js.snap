// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the 'empty-project' fixture returns the correct services: services count 1`] = `0`;

exports[`From within the 'example-todo-main' fixture returns the correct services: /api/src/services/todos/todos.js content 1`] = `
RedwoodService {
  "filepath": "/api/src/services/todos/todos.js",
  "functions": [
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todos",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "createTodo",
      "parameters": [
        "body",
      ],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "numTodos",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "updateTodoStatus",
      "parameters": [
        "id",
        "status",
      ],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "renameTodo",
      "parameters": [
        "id",
        "body",
      ],
      "serviceName": "todos",
      "type": "service-function",
    },
  ],
  "name": "todos",
  "type": "service",
}
`;

exports[`From within the 'example-todo-main' fixture returns the correct services: /api/src/services/todos/todos.js errors 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct services: /api/src/services/todos/todos.js warnings 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct services: services count 1`] = `1`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct services: /api/src/services/todos/todos.js content 1`] = `
RedwoodService {
  "filepath": "/api/src/services/todos/todos.js",
  "functions": [
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todos",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithStringRole",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMultipleRoles",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithInvalidRole",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithInvalidRoles",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMissingRoles",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithBuiltInDeprecatedDirective",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMissingRolesAttribute",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMissingRolesAttributeNumeric",
      "parameters": [],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "createTodo",
      "parameters": [
        "body",
      ],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "updateTodoStatus",
      "parameters": [
        "id",
        "status",
      ],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "renameTodo",
      "parameters": [
        "id",
        "body",
      ],
      "serviceName": "todos",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "deleteTodo",
      "parameters": [
        "id",
      ],
      "serviceName": "todos",
      "type": "service-function",
    },
  ],
  "name": "todos",
  "type": "service",
}
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct services: /api/src/services/todos/todos.js errors 1`] = `[]`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct services: /api/src/services/todos/todos.js warnings 1`] = `[]`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct services: services count 1`] = `1`;

exports[`From within the 'test-project' fixture returns the correct services: /api/src/services/contacts/contacts.ts content 1`] = `
RedwoodService {
  "filepath": "/api/src/services/contacts/contacts.ts",
  "functions": [
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "contacts",
      "parameters": [],
      "serviceName": "contacts",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "contact",
      "parameters": [
        "id",
      ],
      "serviceName": "contacts",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "createContact",
      "parameters": [
        "input",
      ],
      "serviceName": "contacts",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "updateContact",
      "parameters": [
        "id",
        "input",
      ],
      "serviceName": "contacts",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "deleteContact",
      "parameters": [
        "id",
      ],
      "serviceName": "contacts",
      "type": "service-function",
    },
  ],
  "name": "contacts",
  "type": "service",
}
`;

exports[`From within the 'test-project' fixture returns the correct services: /api/src/services/contacts/contacts.ts errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct services: /api/src/services/contacts/contacts.ts warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct services: /api/src/services/posts/posts.ts content 1`] = `
RedwoodService {
  "filepath": "/api/src/services/posts/posts.ts",
  "functions": [
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "posts",
      "parameters": [],
      "serviceName": "posts",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "post",
      "parameters": [
        "id",
      ],
      "serviceName": "posts",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "createPost",
      "parameters": [
        "input",
      ],
      "serviceName": "posts",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "updatePost",
      "parameters": [
        "id",
        "input",
      ],
      "serviceName": "posts",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "deletePost",
      "parameters": [
        "id",
      ],
      "serviceName": "posts",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "Post",
      "parameters": [],
      "serviceName": "posts",
      "type": "service-function",
    },
  ],
  "name": "posts",
  "type": "service",
}
`;

exports[`From within the 'test-project' fixture returns the correct services: /api/src/services/posts/posts.ts errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct services: /api/src/services/posts/posts.ts warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct services: /api/src/services/users/users.ts content 1`] = `
RedwoodService {
  "filepath": "/api/src/services/users/users.ts",
  "functions": [
    RedwoodServiceFunction {
      "constructorErrors": [
        {
          "component": {
            "filepath": "/Users/jgmw/Development/redwood/redwood/__fixtures__/test-project/api/src/services/users/users.ts",
            "name": "users",
            "type": "service-function",
          },
          "message": "Unable to extract the name of the service function",
        },
        {
          "component": {
            "filepath": "/Users/jgmw/Development/redwood/redwood/__fixtures__/test-project/api/src/services/users/users.ts",
            "name": "users",
            "type": "service-function",
          },
          "message": "Unable to extract the parameters of the service function",
        },
      ],
      "filepath": "/api/src/services/users/users.ts",
      "name": "users",
      "parameters": [],
      "serviceName": "users",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/users/users.ts",
      "name": "user",
      "parameters": [
        "id",
      ],
      "serviceName": "users",
      "type": "service-function",
    },
    RedwoodServiceFunction {
      "constructorErrors": [],
      "filepath": "/api/src/services/users/users.ts",
      "name": "User",
      "parameters": [],
      "serviceName": "users",
      "type": "service-function",
    },
  ],
  "name": "users",
  "type": "service",
}
`;

exports[`From within the 'test-project' fixture returns the correct services: /api/src/services/users/users.ts errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct services: /api/src/services/users/users.ts warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct services: services count 1`] = `3`;
